couchbase/endpoint_diagnostics.hxx

```diff
@@ -0,0 +1,204 @@
+     * @committed
+     */
+    connected,
+
+    /**
+     * The endpoint is disconnected but trying to connect right now.
+     *
+     * @since 1.0.0
+     * @committed
+     */
+    connecting,
+
+    /**
+     * The endpoint is disconnected (not reachable) and not trying to connect.
+     *
+     * @since 1.0.0
+     * @committed
+     */
+    disconnected,
+
+    /**
+     * The endpoint is currently disconnecting.
+     *
+     * @since 1.0.0
+     * @committed
+     */
+    disconnecting,
+};
+
+class endpoint_diagnostics
+{
+  public:
+    /**
+     * @since 1.0.0
+     * @internal
+     */
+    endpoint_diagnostics(service_type type,
+                         std::string id,
+                         std::optional<std::chrono::microseconds> last_activity,
+                         std::string local,
+                         std::string remote,
+                         std::optional<std::string> endpoint_namespace,
+                         endpoint_state state,
+                         std::optional<std::string> details)
+      : type_{ type }
+      , id_{ std::move(id) }
+      , last_activity_{ last_activity }
+      , local_{ std::move(local) }
+      , remote_{ std::move(remote) }
+      , namespace_{ std::move(endpoint_namespace) }
+      , state_{ state }
+      , details_{ std::move(details) }
+    {
+    }
+
+    /**
+     * Returns the service type for this endpoint.
+     *
+     * @return the service type.
+     *
+     * @since 1.0.0
+     * @committed
+     */
+    [[nodiscard]] auto type() const -> service_type
+    {
+        return type_;
+    }
+
+    /**
+     *
+     * Returns the ID for this endpoint.
+     *
+     * @return the endpoint ID.
+     *
+     * @since 1.0.0
+     * @committed
+     */
+    [[nodiscard]] auto id() const -> std::string
+    {
+        return id_;
+    }
+
+    /**
+     * Returns the time since the last activity, if there has been one.
+     *
+     * @return the duration since the last activity.
+     *
+     * @since 1.0.0
+     * @committed
+     */
+    [[nodiscard]] auto last_activity() const -> std::optional<std::chrono::microseconds>
+    {
+        return last_activity_;
+    }
+
+    /**
+     * Returns the local socket address for this endpoint.
+     *
+     * @return the local socket address.
+     *
+     * @since 1.0.0
+     * @committed
+     */
+    [[nodiscard]] auto local() const -> std::string
+    {
+        return local_;
+    }
+
+    /**
+     * Returns the remote socket address for this endpoint.
+     *
+     * @return the remote socket address.
+     *
+     * @since 1.0.0
+     * @committed
+     */
+    [[nodiscard]] auto remote() const -> std::string
+    {
+        return remote_;
+    }
+
+    /**
+     * Returns the namespace of this endpoint (likely the bucket name if present).
+     *
+     * @return the namespace.
+     *
+     * @since 1.0.0
+     * @committed
+     */
+    [[nodiscard]] auto endpoint_namespace() const -> std::optional<std::string>
```

Comment:
Since we can't call this `namespace` that the RFC wants, I called it `endpoint_namespace` to work around the fact that it's a reserved keyword in C++. In the core it's called `bucket` which is what it mostly is at the moment, but could it be possible that in the future the namespace is not just the bucket name?